server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java

```diff
@@ -193,50 +196,44 @@ public CacheBackupEntry next() {
-                  return Flowable.fromCompletionStage(cs);
-               }, batchSize)
-               .count();
-
-         return restored
-               .observeOn(Schedulers.from(blockingManager.asExecutor("restore-cache-" + cacheName)))
-               .toCompletionStage()
-               .handle((entries, t) -> {
-                  try {
-                     is.close();
-                  } catch (IOException ignore) { }
-
-                  if (t != null) {
-                     throw CompletableFutures.asCompletionException(t);
-                  }
-
-                  log.debugf("Cache %s restored %d entries", cacheName, entries);
-                  return null;
-               });
-      } catch (IOException e) {
-         throw new CacheException(e);
-      }
+      AdvancedCache<Object, Object> cache = cm.getCache(cacheName).getAdvancedCache();
+      ComponentRegistry cr = SecurityActions.getCacheComponentRegistry(cache);
+      CommandsFactory commandsFactory = cr.getCommandsFactory();
+      KeyPartitioner keyPartitioner = cr.getComponent(KeyPartitioner.class);
+      InvocationHelper invocationHelper = cr.getComponent(InvocationHelper.class);
+      StorageConfigurationManager scm = cr.getComponent(StorageConfigurationManager.class);
+      PersistenceMarshaller persistenceMarshaller = cr.getPersistenceMarshaller();
+      Marshaller userMarshaller = persistenceMarshaller.getUserMarshaller();
+
+      boolean keyMarshalling = !scm.getKeyStorageMediaType().isBinary();
+      boolean valueMarshalling = !scm.getValueStorageMediaType().isBinary();
+
+      int batchSize = SecurityActions.getCacheConfiguration(cache).clustering().stateTransfer().chunkSize();
+      return Flowable.fromPublisher(blockingManager.blockingPublisher(f))
+            .rebatchRequests(batchSize)
+            .map(entry -> {
+               Object key = keyMarshalling ? unmarshall(entry.key, userMarshaller) : scm.getKeyWrapper().wrap(entry.key);
+               Object value = valueMarshalling ? unmarshall(entry.value, userMarshaller) : scm.getValueWrapper().wrap(entry.value);
+               Metadata metadata = unmarshall(entry.metadata, persistenceMarshaller);
+               Metadata internalMetadataImpl = new InternalMetadataImpl(metadata, entry.created, entry.lastUsed);
+
+               PutKeyValueCommand cmd = commandsFactory.buildPutKeyValueCommand(key, value, keyPartitioner.getSegment(key),
+                     internalMetadataImpl, FlagBitSets.IGNORE_RETURN_VALUES);
+               commandsFactory.buildPutKeyValueCommand(key, value, keyPartitioner.getSegment(key),
+                     internalMetadataImpl, FlagBitSets.IGNORE_RETURN_VALUES);
+               cmd.setInternalMetadata(entry.internalMetadata);
+               return cmd;
+            })
+            .flatMap(cmd -> RxJavaInterop.voidCompletionStageToFlowable(invocationHelper.invokeAsync(cmd, 1)), batchSize)
```

Comment:
Just a tip, but this internally has a parallelism to limit how many outstanding operations (in this case put commands) can be done.  The default is Flowable.bufferSize which is 128. I _think_ that should be fine for this case as this is not a CPU based parallelism anyways and is limited by network/disk instead.